src/main/java/com/coreos/jetcd/EtcdLeaseImpl.java

```diff
@@ -0,0 +1,300 @@
+            long now = System.currentTimeMillis();
+            List<Long> expireLeases = new ArrayList<>();
+            for(Lease l : keepAlives.values()){
+                if(now > l.getDeadLine()){
+                    expireLeases.add(l.getLeaseID());
+                }
+            }
+
+            for(Long id: expireLeases){
+                removeLease(id);
+                if(leaseHandler!=null){
+                    leaseHandler.onLeaseExpired(id);
+                }
+            }
+        }
+    }
+
+    /**
+     * This method update the deadline and NextKeepAlive time
+     * @param leaseKeepAliveResponse The response receive from etcd server
+     */
+    public void processKeepLiveRespond(LeaseKeepAliveResponse leaseKeepAliveResponse){
+        synchronized (lock){
+            if(keepAlives.containsKey(leaseKeepAliveResponse.getID())){
+
+                if(leaseKeepAliveResponse.getTTL() <= 0){
+                    removeLease(leaseKeepAliveResponse.getID());
+                    if(leaseHandler!=null){
+                        leaseHandler.onLeaseExpired(leaseKeepAliveResponse.getID());
+                    }
+                }else{
+                    Lease lease = keepAlives.get(leaseKeepAliveResponse.getID());
+                    long nextKeepAlive =
+                            System.currentTimeMillis() + 1000 + leaseKeepAliveResponse.getTTL() * 1000 / 3;
+                    lease.setNextKeepAlive(nextKeepAlive);
+                    lease.setDeadLine(System.currentTimeMillis() + leaseKeepAliveResponse.getTTL() * 1000);
+
+                }
+            }
+        }
+    }
+
+    /**
+     * remove the lease from keep alive map
+     * @param leaseId
+     */
+    private void removeLease(long leaseId){
+        synchronized (lock){
+            if(keepAlives.containsKey(leaseId)){
+                keepAlives.remove(leaseId);
+            }
+        }
+    }
+
+    private LeaseKeepAliveRequest newKeepAliveRequest(long leaseId){
+        return LeaseKeepAliveRequest.newBuilder()
+                        .setID(leaseId)
+                        .build();
+    }
+
+    private void newRequestStream(){
+        synchronized (lock){
+            if(keepAliveRequestStreamObserver!=null){
+                keepAliveRequestStreamObserver.onCompleted();
+                keepAliveRequestStreamObserver=null;
+            }
+            keepAliveRequestStreamObserver = leaseStub.leaseKeepAlive(keepAliveResponseStreamObserver);
+        }
+    }
+
+    /**
+     * close all the keep alive thread and process
+     */
+    public void close(){
+        synchronized (lock){
+            if(keepAliveRequestStreamObserver!=null){
+                keepAliveRequestStreamObserver.onCompleted();
+                keepAliveRequestStreamObserver = null;
+            }
+        }
+    }
+
+    public  interface EtcdLeaseHandler{
+        void onLeaseExpired(long leaseId);
+        void onThrowable(Throwable throwable);
```

Comment:
Ok, I will modfiy its name. I want to add a method named onResponse to give user response, what do you think about.